python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,601 @@
+        return None
+
+    @to.register(IntegerType)
+    def _(self, type_var):
+        return self
+
+    @to.register(LongType)
+    def _(self, type_var):
+        return LongLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(DateType)
+    def _(self, type_var):
+        return DateLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(
+                Decimal(self.value).quantize(
+                    Decimal("." + "".join(["0" for _ in range(1, type_var.scale)]) + "1"), rounding=ROUND_HALF_UP
+                )
+            )
+
+
+class LongLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(LongLiteral, self).__init__(f"LongLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(LongType)
+    def _(self, type_var):
+        return self
+
+    @to.register(IntegerType)
+    def _(self, type_var):
+        if IntegerType.max < self.value:
+            return AboveMax()
+        elif IntegerType.min > self.value:
+            return BelowMin()
+        return IntegerLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(TimeType)
+    def _(self, type_var):
+        return TimeLiteral(self.value)
+
+    @to.register(TimestampType)
+    def _(self, type_var):
+        return TimestampLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class FloatLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(FloatLiteral, self).__init__(f"FloatLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return self
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal("1."), rounding=ROUND_HALF_UP))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class DoubleLiteral(ComparableLiteral):
+    def __init__(self, value):
```

Comment:
I think the constructors should have type annotations so that we correctly convert values when converting literals.